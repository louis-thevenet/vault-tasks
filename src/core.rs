use color_eyre::{eyre::bail, Result};
use serde::Deserialize;
use task::Task;

use std::{
    collections::HashSet,
    fmt::Display,
    path::{Path, PathBuf},
};
use vault_data::VaultData;

use filter::{filter, Filter};
use tracing::{debug, error};
use vault_parser::VaultParser;

pub mod filter;
pub mod parser;
pub mod sorter;
pub mod task;
pub mod vault_data;
mod vault_parser;

#[derive(Clone, Debug, Deserialize)]
pub struct TaskMarkerConfig {
    pub done: char,
    pub todo: char,
    pub incomplete: char,
    pub canceled: char,
}

// Mostly for tests
impl Default for TaskMarkerConfig {
    fn default() -> Self {
        Self {
            done: 'x',
            todo: ' ',
            incomplete: '/',
            canceled: '-',
        }
    }
}

#[derive(Clone, Debug, Deserialize)]
pub struct PrettySymbolsConfig {
    #[serde(default)]
    pub task_done: String,
    #[serde(default)]
    pub task_todo: String,
    #[serde(default)]
    pub task_incomplete: String,
    #[serde(default)]
    pub task_canceled: String,
    #[serde(default)]
    pub due_date: String,
    #[serde(default)]
    pub priority: String,
    #[serde(default)]
    pub today_tag: String,
    #[serde(default)]
    pub progress_bar_true: String,
    #[serde(default)]
    pub progress_bar_false: String,
}
impl Default for PrettySymbolsConfig {
    fn default() -> Self {
        Self {
            task_done: String::from("✅"),
            task_todo: String::from("❌"),
            task_incomplete: String::from("⏳"),
            task_canceled: String::from("🚫"),
            due_date: String::from("📅"),
            priority: String::from("❗"),
            today_tag: String::from("☀️"),
            progress_bar_true: String::from("🟩"),
            progress_bar_false: String::from("⬜️"),
        }
    }
}
#[derive(Clone, Debug, Deserialize, Default)]
pub struct TasksConfig {
    #[serde(default)]
    pub parse_dot_files: bool,
    #[serde(default)]
    pub file_tags_propagation: bool,
    #[serde(default)]
    pub ignored: Vec<PathBuf>,
    #[serde(default)]
    pub indent_length: usize,
    #[serde(default)]
    pub use_american_format: bool,
    #[serde(default)]
    pub show_relative_due_dates: bool,
    #[serde(default)]
    pub completion_bar_length: usize,
    #[serde(default)]
    pub vault_path: PathBuf,
    #[serde(default)]
    pub tasks_drop_file: String,
    #[serde(default)]
    pub explorer_default_search_string: String,
    #[serde(default)]
    pub filter_default_search_string: String,
    #[serde(default)]
    pub task_state_markers: TaskMarkerConfig,
    #[serde(default)]
    pub pretty_symbols: PrettySymbolsConfig,
}

pub struct TaskManager {
    pub tasks: VaultData,
    config: TasksConfig,
    pub tags: HashSet<String>,
    pub current_filter: Option<Filter>,
}
impl Default for TaskManager {
    fn default() -> Self {
        Self {
            tasks: VaultData::Directory("Empty Vault".to_owned(), vec![]),
            tags: HashSet::new(),
            current_filter: None,
            config: TasksConfig::default(),
        }
    }
}
impl TaskManager {
    /// Loads a vault from a `Config` and returns a `TaskManager`.
    ///
    /// # Errors
    ///
    /// This function will return an error if the vault can't be loaded.
    pub fn load_from_config(config: &TasksConfig) -> Result<Self> {
        let mut res = Self::default();
        res.reload(config)?;
        Ok(res)
    }

    /// Reloads the `VaultData` from file system.
    ///
    /// # Errors
    ///
    /// This function will return an error if the vault can't be parsed, or if tasks can't be fixed (relative dates are replaced by fixed dates for example).
    pub fn reload(&mut self, config: &TasksConfig) -> Result<()> {
        self.config = config.clone();
        let vault_parser = VaultParser::new(config.clone());
        let tasks = vault_parser.scan_vault()?;

        Self::rewrite_vault_tasks(config, &tasks)
            .unwrap_or_else(|e| error!("Failed to fix tasks: {e}"));

        let mut tags = HashSet::new();
        Self::collect_tags(&tasks, &mut tags);

        self.tasks = tasks;
        self.tags = tags;
        Ok(())
    }

    /// Explores the vault and fills a `&mut HashSet<String>` with every tags found.
    pub fn collect_tags(tasks: &VaultData, tags: &mut HashSet<String>) {
        match tasks {
            VaultData::Directory(_, children) | VaultData::Header(_, _, children) => {
                children.iter().for_each(|c| Self::collect_tags(c, tags));
            }
            VaultData::Task(task) => {
                task.tags.clone().unwrap_or_default().iter().for_each(|t| {
                    tags.insert(t.clone());
                });
                task.subtasks
                    .iter()
                    .for_each(|task| Self::collect_tags(&VaultData::Task(task.clone()), tags));
            }
        }
    }

    /// Follows the `selected_header_path` to retrieve the correct `VaultData`.
    /// Then returns every `VaultData` objects on the same layer.
    ///
    /// # Errors
    /// Will return an error if the vault is empty or the first layer is not a `VaultData::Directory`
    pub fn get_explorer_entries(&self, selected_header_path: &[String]) -> Result<Vec<VaultData>> {
        fn aux(
            file_entry: Vec<VaultData>,
            selected_header_path: &[String],
            path_index: usize,
        ) -> Result<Vec<VaultData>> {
            if path_index == selected_header_path.len() {
                Ok(file_entry)
            } else {
                for entry in file_entry {
                    match entry {
                        VaultData::Directory(name, children)
                        | VaultData::Header(_, name, children) => {
                            if name == selected_header_path[path_index] {
                                return aux(children, selected_header_path, path_index + 1);
                            }
                        }
                        VaultData::Task(task) => {
                            if task.name == selected_header_path[path_index] {
                                return aux(
                                    task.subtasks
                                        .iter()
                                        .map(|t| VaultData::Task(t.clone()))
                                        .collect(),
                                    selected_header_path,
                                    path_index + 1,
                                );
                            }
                        }
                    }
                }
                bail!("Couldn't find corresponding entry");
            }
        }

        let filtered_tasks = if let Some(task_filter) = &self.current_filter {
            filter(&self.tasks, task_filter)
        } else {
            Some(self.tasks.clone())
        };

        match filtered_tasks {
            Some(VaultData::Directory(_, entries)) => aux(entries, selected_header_path, 0),
            None => bail!("Empty Vault"),
            _ => {
                error!("First layer of VaultData was not a Directory");
                bail!("First layer of VaultData was not a Directory")
            }
        }
    }

    /// Same as `get_explorer_entries`, but discards any children of the entries.
    ///
    /// # Errors
    ///
    /// This function will return an error if the path can't be resolved.
    pub fn get_explorer_entries_without_children(&self, path: &[String]) -> Result<Vec<VaultData>> {
        Ok(self
            .get_explorer_entries(path)? // Get the entries at the path
            .iter() // Discard every children
            .map(|vd| match vd {
                VaultData::Directory(name, _) => VaultData::Directory(name.clone(), vec![]),
                VaultData::Header(level, name, _) => {
                    VaultData::Header(*level, name.clone(), vec![])
                }
                VaultData::Task(t) => {
                    let mut t = t.clone();
                    t.subtasks = vec![]; // Discard subtasks
                    VaultData::Task(t)
                }
            })
            .collect::<Vec<VaultData>>())
    }

    /// Recursively calls `Task.fix_task_attributes` on every task from the vault.
    fn rewrite_vault_tasks(config: &TasksConfig, tasks: &VaultData) -> Result<()> {
        fn explore_tasks_rec(
            config: &TasksConfig,
            filename: &mut PathBuf,
            file_entry: &VaultData,
        ) -> Result<()> {
            match file_entry {
                VaultData::Header(level, name, children) => {
                    let mut filename = filename.clone();
                    if *level == 0 {
                        filename.push(name);
                    }
                    children
                        .iter()
                        .try_for_each(|c| explore_tasks_rec(config, &mut filename, c))?;
                }
                VaultData::Task(task) => {
                    task.fix_task_attributes(config, filename)?;
                    task.subtasks
                        .iter()
                        .try_for_each(|t| t.fix_task_attributes(config, filename))?;
                }
                VaultData::Directory(dir_name, children) => {
                    let mut filename = filename.clone();
                    filename.push(dir_name);
                    children
                        .iter()
                        .try_for_each(|c| explore_tasks_rec(config, &mut filename.clone(), c))?;
                }
            }
            Ok(())
        }
        explore_tasks_rec(config, &mut PathBuf::new(), tasks)
    }
    /// Retrieves the `VaultData` at the given `path`, and returns the entries to display.
    ///
    /// If the path ends with a task, the `task_preview_offset` parameter determines whether the function should return the task itself or its content (subtasks) as with directories and headers.
    pub fn get_vault_data_from_path(&self, path: &[String]) -> Result<VaultData> {
        /// Recursively searches for the entry in the vault.
        /// `path_index` is the index of the current path element we are looking for.
        fn aux(
            file_entry: VaultData,
            selected_header_path: &[String],
            path_index: usize,
        ) -> Result<VaultData> {
            // Remaining path is empty?
            if path_index == selected_header_path.len() {
                Ok(file_entry)
            } else {
                match &file_entry {
                    // Both variants are very similar
                    VaultData::Header(_, name, children) | VaultData::Directory(name, children) => {
                        if *name == selected_header_path[path_index] {
                            if path_index + 1 == selected_header_path.len() {
                                return Ok(file_entry.clone());
                            }
                            // Look for the child that matches the path
                            for child in children {
                                if let Ok(found) =
                                    aux(child.clone(), selected_header_path, path_index + 1)
                                {
                                    return Ok(found);
                                    // I'm tempted to break here but we might have multiple entries with the same name
                                }
                            }
                        }
                        // Either it's the first layer and the path is wrong or we recursively called on the wrong entry which is impossible
                        bail!("Couldn't find corresponding entry");
                    }
                    VaultData::Task(task) => {
                        if task.name == selected_header_path[path_index] {
                            // If we are at the end of the path, we return the task itself
                            // This depends on the `task_preview_offset` parameter
                            if path_index + 1 == selected_header_path.len() {
                                return Ok(file_entry.clone());
                            }
                            for child in &task.subtasks {
                                if let Ok(found) = aux(
                                    VaultData::Task(child.clone()),
                                    selected_header_path,
                                    path_index + 1,
                                ) {
                                    return Ok(found);
                                }
                            }
                        }
                        bail!("Couldn't find corresponding entry");
                    }
                }
            }
        }

        let filtered_tasks = if let Some(task_filter) = &self.current_filter {
            filter(&self.tasks, task_filter)
        } else {
            Some(self.tasks.clone())
        };
        match filtered_tasks {
            Some(VaultData::Directory(_, entries)) => {
                for entry in entries {
                    if let Ok(res) = aux(entry, path, 0) {
                        return Ok(res);
                    }
                }
                error!("Vault was not empty but the entry was not found");
                bail!("Vault was not empty but the entry was not found");
            }
            None => bail!("Empty Vault"),
            _ => {
                error!("First layer of VaultData was not a Directory");
                bail!("Empty Vault")
            }
        }
    }

    /// Whether the path resolves to something that can be entered or not.
    /// Directories, Headers and Tasks with subtasks can be entered.
    #[must_use]
    pub fn can_enter(&self, selected_header_path: &[String]) -> bool {
        fn aux(file_entry: VaultData, selected_header_path: &[String], path_index: usize) -> bool {
            if path_index == selected_header_path.len() {
                true
            } else {
                match file_entry {
                    VaultData::Directory(name, children) | VaultData::Header(_, name, children) => {
                        if name == selected_header_path[path_index] {
                            return children
                                .iter()
                                .any(|c| aux(c.clone(), selected_header_path, path_index + 1));
                        }
                        false
                    }
                    VaultData::Task(task) => {
                        if task.name == selected_header_path[path_index] {
                            return task.subtasks.iter().any(|t| {
                                aux(
                                    VaultData::Task(t.clone()),
                                    selected_header_path,
                                    path_index + 1,
                                )
                            });
                        }
                        false
                    }
                }
            }
        }

        let filtered_tasks = if let Some(task_filter) = &self.current_filter {
            filter(&self.tasks, task_filter)
        } else {
            return false;
        };
        let Some(VaultData::Directory(_, entries)) = filtered_tasks else {
            return false;
        };
        entries
            .iter()
            .any(|e| aux(e.clone(), selected_header_path, 0))
    }

    pub fn add_task(&mut self, task_desc: &str, filename_opt: Option<String>) {
        /// Recursively searches for the entry in the vault.
        /// `path_index` is the index of the current path element we are looking for.
        fn aux(file_entry: &mut VaultData, filename: &str, new_task: &Task) -> Result<()> {
            match file_entry {
                VaultData::Directory(name, children) => {
                    // Look for the child that matches the path
                    for child in children {
                        if let Ok(()) = aux(child, filename, new_task) {
                            return Ok(());
                            // I'm tempted to break here but we might have multiple entries with the same name
                        }
                    }
                    // Either it's the first layer and the path is wrong or we recursively called on the wrong entry which is impossible
                    bail!("Couldn't find corresponding entry in Directory {name}");
                }
                VaultData::Header(_, name, children) => {
                    if *name == filename {
                        debug! {"Adding task to {name}"
                        };

                        children.push(VaultData::Task(new_task.clone()));
                        return Ok(());
                    }
                    // Look for the child that matches the path
                    for child in children {
                        if let Ok(()) = aux(child, filename, new_task) {
                            return Ok(());
                            // I'm tempted to break here but we might have multiple entries with the same name
                        }
                    }
                    // Either it's the first layer and the path is wrong or we recursively called on the wrong entry which is impossible
                    bail!("Couldn't find corresponding entry in Header {name}");
                }

                VaultData::Task(_task) => {
                    bail!("Adding subtasks from CLI is not supported yet");
                }
            }
        }

        // Get filename
        let filename = &filename_opt.unwrap_or(self.config.tasks_drop_file.clone());
        if filename.is_empty() {
            eprintln!( "No drop file was provided via `--filename`, and no default is set in the configuration." );
            return;
        }
        if !Path::new(&filename)
            .extension()
            .is_some_and(|ext| ext.eq_ignore_ascii_case("md"))
        {
            eprintln!("Filename {filename} does not have the .md extension");
            return;
        }

        // Parse input string
        let vault_parser = VaultParser::new(self.config.clone());
        let task = match vault_parser.parse_single_task(task_desc, filename) {
            Ok(task) => task,
            Err(e) => {
                eprintln!("Failed to parse task: {e}");
                return;
            }
        };
        debug!("Adding new task: {} to path: {:?}", task, filename);

        // Insert the task into the vault tree
        if let Err(e) = aux(&mut self.tasks, filename, &task) {
            eprintln!("Failed to insert task in VaultData tree: {e}");
            return;
        }

        // Fix attributes again (maybe we should only fix the task itself
        // but we would need the path to filename)
        if let Err(e) = Self::rewrite_vault_tasks(&self.config, &self.tasks) {
            eprintln!("Failed to fix task attributes in vault files: {e}");
        }
    }
}
impl Display for TaskManager {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.tasks)?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashSet;

    use super::TaskManager;

    use crate::core::{task::Task, vault_data::VaultData};

    #[test]
    fn test_get_vault_data() {
        let expected_tasks = vec![
            VaultData::Task(Task {
                name: "test".to_string(),
                line_number: Some(8),
                description: Some("test\ndesc".to_string()),
                ..Default::default()
            }),
            VaultData::Task(Task {
                name: "test".to_string(),
                line_number: Some(8),
                description: Some("test\ndesc".to_string()),
                ..Default::default()
            }),
            VaultData::Task(Task {
                name: "test".to_string(),
                line_number: Some(8),
                description: Some("test\ndesc".to_string()),
                ..Default::default()
            }),
        ];
        let expected_header = VaultData::Header(2, "2".to_string(), vec![]);
        let input = VaultData::Directory(
            "test".to_owned(),
            vec![VaultData::Header(
                0,
                "Test".to_string(),
                vec![
                    VaultData::Header(
                        1,
                        "1".to_string(),
                        vec![VaultData::Header(2, "2".to_string(), vec![])],
                    ),
                    VaultData::Header(
                        1,
                        "1.2".to_string(),
                        vec![
                            VaultData::Header(3, "3".to_string(), vec![]),
                            VaultData::Header(2, "4".to_string(), expected_tasks.clone()),
                        ],
                    ),
                ],
            )],
        );

        let task_mgr = TaskManager {
            tasks: input,
            tags: HashSet::new(),
            ..Default::default()
        };

        let path = vec![String::from("Test"), String::from("1"), String::from("2")];
        let res = task_mgr.get_vault_data_from_path(&path).unwrap();
        assert_eq!(expected_header, res);

        let path = vec![String::from("Test"), String::from("1.2"), String::from("4")];
        let res = task_mgr.get_vault_data_from_path(&path).unwrap();
        let expected_header = VaultData::Header(2, "4".to_string(), expected_tasks.clone());
        assert_eq!(expected_header, res);
    }
}
