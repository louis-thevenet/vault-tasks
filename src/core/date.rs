use std::{cmp::Ordering, fmt::Display};

use chrono::{NaiveDate, NaiveDateTime, Timelike};

#[derive(Debug, Hash, Eq, PartialEq, Clone)]
/// This type accounts for the case where the task has a due date but no exact due time
pub enum DueDate {
    NoDate,
    Day(NaiveDate),
    DayTime(NaiveDateTime),
}
impl Display for DueDate {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Day(date) => write!(f, "{date}"),
            Self::DayTime(date) => write!(f, "{date}"),
            Self::NoDate => Ok(()),
        }
    }
}

impl DueDate {
    #[must_use]
    pub fn to_display_format(&self, due_date_symbol: &str, not_american_format: bool) -> String {
        if matches!(self, Self::NoDate) {
            String::new()
        } else {
            format!(
                "{due_date_symbol} {}",
                self.to_string_format(not_american_format)
            )
        }
    }
    #[must_use]
    pub fn to_string_format(&self, not_american_format: bool) -> String {
        let format_date = if not_american_format {
            "%d/%m/%Y"
        } else {
            "%Y/%m/%d"
        };
        let format_datetime = if not_american_format {
            "%d/%m/%Y %T"
        } else {
            "%Y/%m/%d %T"
        };

        match self {
            Self::Day(date) => date.format(format_date).to_string(),
            Self::DayTime(date) => date.format(format_datetime).to_string(),
            Self::NoDate => String::new(),
        }
    }

    #[must_use]
    pub fn get_relative_str(&self) -> Option<String> {
        // This truncation prevents errors such as 23:59:59:999... instead of 24 hours
        let now = chrono::Local::now()
            .with_second(0)
            .unwrap()
            .with_nanosecond(0)
            .unwrap();

        let time_delta = match self {
            DueDate::Day(naive_date) => now.date_naive().signed_duration_since(*naive_date),
            DueDate::DayTime(naive_date_time) => {
                now.date_naive()
                    .signed_duration_since(naive_date_time.date())
                    // same truncation here
                    + now.time().signed_duration_since(
                        naive_date_time
                            .time()
                            .with_second(0)
                            .unwrap()
                            .with_nanosecond(0)
                            .unwrap(),
                    )
            }
            DueDate::NoDate => return None,
        };
        let (prefix, suffix) = match time_delta.num_seconds().cmp(&0) {
            Ordering::Less => (String::from("in "), String::new()),
            Ordering::Equal => (String::new(), String::new()),
            Ordering::Greater => (String::new(), String::from(" ago")),
        };

        let time_delta_abs = time_delta.abs();

        if time_delta_abs.is_zero() {
            return Some(String::from("today"));
        }
        if time_delta.num_seconds() < 0 && time_delta_abs.num_hours() == 24 {
            return Some(String::from("tomorrow"));
        }
        if time_delta.num_seconds() > 0 && time_delta_abs.num_hours() == 24 {
            return Some(String::from("yesterday"));
        }

        // >= 13 months -> show years
        let res = if 4 * 12 < time_delta_abs.num_weeks() {
            format!("{} years", time_delta_abs.num_weeks() / (12 * 4))
            // >= 5 weeks -> show months
        } else if 5 <= time_delta_abs.num_weeks() {
            format!("{} months", time_delta_abs.num_weeks() / 4)
            // >= 2 weeks -> show weeks
        } else if 2 <= time_delta_abs.num_weeks() {
            format!("{} weeks", time_delta_abs.num_weeks())
            // >= 2 days -> show days
        } else if 2 <= time_delta_abs.num_days() {
            format!("{} days", time_delta_abs.num_days())
            // >= 2 hours -> show hours
        } else if 2 <= time_delta_abs.num_hours() {
            format!("{} hours", time_delta_abs.num_hours())
        } else {
            format!("{} minutes", time_delta_abs.num_minutes())
        };
        Some(format!("{prefix}{res}{suffix}"))
    }
}
