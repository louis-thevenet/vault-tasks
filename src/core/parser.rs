mod parser_date;
#[allow(clippy::module_name_repetitions)]
pub mod parser_file_entry;
mod parser_time;
pub mod task;
pub mod tracker;
