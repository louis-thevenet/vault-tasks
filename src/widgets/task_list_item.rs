use std::rc::Rc;

use ratatui::{
    prelude::*,
    widgets::{Block, Borders, Paragraph},
};
use ratskin::RatSkin;
use tracing::error;

use crate::core::{PrettySymbolsConfig, task::Task, vault_data::VaultData};

const HEADER_INDENT_RATIO: u16 = 3;

#[derive(Clone)]
pub struct TaskListItem {
    item: VaultData,
    pub height: u16,
    symbols: PrettySymbolsConfig,
    not_american_format: bool,
    completion_bar_length: usize,
    show_relative_due_dates: bool,
    max_width: u16,
    display_filename: bool,
    header_style: Style,
}

impl TaskListItem {
    pub fn header_style(mut self, style: Style) -> Self {
        self.header_style = style;
        self
    }
    pub fn new(
        item: VaultData,
        not_american_format: bool,
        symbols: PrettySymbolsConfig,
        max_width: u16,
        display_filename: bool,
        show_relative_due_dates: bool,

        completion_bar_length: usize,
    ) -> Self {
        let height = Self::compute_height(&item, max_width);
        Self {
            item,
            height,
            not_american_format,
            max_width,
            display_filename,
            symbols,
            header_style: Style::default(),
            show_relative_due_dates,
            completion_bar_length,
        }
    }
    #[allow(clippy::too_many_lines)]
    fn task_to_paragraph(&self, area: Rect, task: &Task) -> (Rc<[Rect]>, Paragraph<'_>) {
        let mut lines = vec![];
        let mut data_line = vec![];

        let rat_skin = RatSkin::default();

        let state = task.state.display(self.symbols.clone());
        let title = state.clone() + " " + &task.name;
        let title_parsed = rat_skin.parse(RatSkin::parse_text(&title), self.max_width);
        let binding = Line::raw(state);
        let title = match title_parsed.first() {
            Some(t) => {
                lines.append(&mut title_parsed[1..].to_vec());
                t
            }
            None => &binding,
        };

        let surrounding_block =
            Block::default()
                .borders(Borders::ALL)
                .title_bottom(if self.display_filename {
                    Line::from(task.filename.clone()).right_aligned()
                } else {
                    Line::from("")
                });

        if task.is_today {
            data_line.push(Span::raw(format!("{} ", self.symbols.today_tag)));
        }

        if let Some(due_date) = &task.due_date {
            data_line.push(Span::from(format!(
                "{} ",
                due_date.to_display_format(&self.symbols.due_date, self.not_american_format)
            )));
            if self.show_relative_due_dates {
                if let Some(due_date_relative) = due_date.get_relative_str() {
                    data_line.push(Span::styled(
                        format!("({due_date_relative}) "),
                        Style::new().dim(),
                    ));
                }
            }
        }
        if let Some(bar) = task.get_completion_bar(
            self.completion_bar_length,
            &(
                self.symbols.progress_bar_false.clone(),
                self.symbols.progress_bar_true.clone(),
            ),
        ) {
            data_line.push(Span::raw(bar));
        }
        if task.priority > 0 {
            data_line.push(Span::raw(format!(
                "{}{} ",
                self.symbols.priority, task.priority
            )));
        }
        if !data_line.is_empty() {
            lines.push(Line::from(data_line));
        }
        let mut tag_line = String::new();
        if task.tags.is_some() {
            tag_line.push_str(
                &task
                    .tags
                    .clone()
                    .unwrap()
                    .iter()
                    .map(|t| format!("#{t}"))
                    .collect::<Vec<String>>()
                    .join(" "),
            );
        }
        if !tag_line.is_empty() {
            lines.push(Line::from(Span::styled(tag_line, Color::DarkGray)));
        }
        if let Some(description) = task.description.clone() {
            let text = rat_skin.parse(RatSkin::parse_text(&description), self.max_width);
            lines = [lines, text].concat();
        }
        let mut constraints = vec![Constraint::Length((lines.len()).try_into().unwrap())];

        for st in &task.subtasks {
            constraints.push(Constraint::Length(Self::compute_height(
                &VaultData::Task(st.clone()),
                self.max_width - 2, // -2 for borders
            )));
        }

        let layout = Layout::default()
            .direction(Direction::Vertical)
            .constraints(constraints)
            .split(surrounding_block.inner(area));

        (
            layout,
            if lines.is_empty() && task.subtasks.is_empty() {
                Paragraph::new(title.clone()).block(surrounding_block)
            } else {
                Paragraph::new(Text::from(lines)).block(surrounding_block.title_top(title.clone()))
            },
        )
    }
    fn compute_height(item: &VaultData, max_width: u16) -> u16 {
        let rat_skin = RatSkin::default();
        match &item {
            VaultData::Directory(_, _) => 1,
            VaultData::Header(_, _, children) => {
                children
                    .iter()
                    .map(|c| Self::compute_height(c, max_width * (100 - HEADER_INDENT_RATIO) / 100))
                    .sum::<u16>()
                    + 1 // name in block (border only on top)
            }
            VaultData::Task(task) => {
                let mut count: u16 = 2; // block
                if 2 + task.name.len() >= max_width as usize {
                    count += (2 + task.name.len() as u16) / max_width;
                }
                if let Some(d) = &task.description {
                    let parsed_desc = rat_skin.parse(RatSkin::parse_text(d), max_width);
                    count += u16::try_from(parsed_desc.len()).unwrap_or_else(|e| {
                        error!("Could not convert description length to u16 :{e}");
                        0
                    });
                }
                if task.due_date.is_some()
                    || task.priority > 0
                    || task.is_today
                    || task.completion.is_some()
                {
                    count += 1;
                }
                if task.tags.is_some() {
                    count += 1;
                }
                for sb in &task.subtasks {
                    count += Self::compute_height(&VaultData::Task(sb.clone()), max_width - 2);
                }
                count.max(3) // If count == 2 then task name will go directly inside a block
                // Else task name will be the block's title and content will go inside
            }
        }
    }
}
impl Widget for TaskListItem {
    fn render(self, area: Rect, buf: &mut Buffer)
    where
        Self: Sized,
    {
        let rat_skin = RatSkin::default();
        match &self.item {
            VaultData::Directory(name, _) => error!("TaskList widget received a directory: {name}"),
            VaultData::Header(_level, name, children) => {
                let surrounding_block = Block::default().borders(Borders::TOP).title(
                    rat_skin
                        .parse(RatSkin::parse_text(name), area.width)
                        .first()
                        .unwrap()
                        .clone()
                        .style(self.header_style),
                );

                let indent = Layout::new(
                    Direction::Horizontal,
                    vec![
                        Constraint::Percentage(HEADER_INDENT_RATIO),
                        Constraint::Percentage(100 - HEADER_INDENT_RATIO),
                    ],
                )
                .split(area);

                let mut constraints = vec![];
                for child in children {
                    constraints.push(Constraint::Length(Self::compute_height(
                        child,
                        self.max_width - indent[0].width,
                    )));
                }
                let layout = Layout::default()
                    .direction(Direction::Vertical)
                    .constraints(constraints)
                    .split(surrounding_block.inner(indent[1]));
                surrounding_block.render(area, buf);

                for (i, child) in children.iter().enumerate() {
                    let sb_widget = Self::new(
                        child.clone(),
                        self.not_american_format,
                        self.symbols.clone(),
                        self.max_width - indent[0].width,
                        self.display_filename,
                        self.show_relative_due_dates,
                        self.completion_bar_length,
                    )
                    .header_style(self.header_style);
                    sb_widget.render(layout[i], buf);
                }
            }
            VaultData::Task(task) => {
                let (layout, par) = self.task_to_paragraph(area, task);
                par.render(area, buf);

                for (i, sb) in task.subtasks.iter().enumerate() {
                    let sb_widget = Self::new(
                        VaultData::Task(sb.clone()),
                        self.not_american_format,
                        self.symbols.clone(),
                        self.max_width - 2, // surrounding block
                        false,
                        self.show_relative_due_dates,
                        self.completion_bar_length,
                    )
                    .header_style(self.header_style);

                    sb_widget.render(layout[i + 1], buf);
                }
            }
        };
    }
}

#[cfg(test)]
mod tests {
    use crate::{
        core::{
            date::Date,
            task::{State, Task},
            vault_data::VaultData,
        },
        widgets::task_list_item::TaskListItem,
    };
    use chrono::NaiveDate;
    use insta::assert_snapshot;
    use ratatui::{Terminal, backend::TestBackend};

    use crate::config::Config;

    #[test]
    fn test_task_list_item() {
        let test_task = VaultData::Task(Task {
            name: "task with a very long title that should wrap to the next line".to_string(),
            state: State::Done,
            tags: Some(vec![String::from("tag"), String::from("tag2")]),
            priority: 5,
            completion: Some(60),
            due_date: Some(Date::DayTime(
                NaiveDate::from_ymd_opt(2016, 7, 8)
                    .unwrap()
                    .and_hms_opt(9, 10, 11)
                    .unwrap(),
            )),
            subtasks: vec![
                Task {
                    name: "subtask with another long title that should wrap around".to_string(),
                    description: Some("test\ndesc".to_string()),
                    ..Default::default()
                },
                Task {
                    name: "subtask test".to_string(),
                    tags: Some(vec![String::from("tag"), String::from("tag2")]),
                    ..Default::default()
                },
                Task {
                    name: "subtask test with a long title 123456789 1 2 3".to_string(),
                    priority: 5,
                    due_date: Some(Date::DayTime(
                        NaiveDate::from_ymd_opt(2016, 7, 8)
                            .unwrap()
                            .and_hms_opt(9, 10, 11)
                            .unwrap(),
                    )),
                    description: Some("test\ndesc".to_string()),
                    ..Default::default()
                },
            ],
            ..Default::default()
        });
        let mut config = Config::default();

        // We don't want tests to be time dependent
        config.tasks_config.show_relative_due_dates = false;

        let max_width = 50;
        let task_list_item = TaskListItem::new(
            test_task,
            false,
            config.tasks_config.pretty_symbols,
            max_width,
            false,
            false,
            config.tasks_config.completion_bar_length,
        );
        let mut terminal = Terminal::new(TestBackend::new(max_width, 40)).unwrap();
        terminal
            .draw(|frame| {
                frame.render_widget(task_list_item, frame.area());
            })
            .unwrap();
        assert_snapshot!(terminal.backend());
    }
}
