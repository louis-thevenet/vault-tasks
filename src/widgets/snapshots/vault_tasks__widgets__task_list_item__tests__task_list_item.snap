---
source: src/widgets/task_list_item.rs
expression: terminal.backend()
---
"┌✅ task with a very long title that ──┐" Hidden by multi-width symbols: [(2, " ")]
"│should wrap to the next line          │"
"│📅 2016/07/08 09:10:11 ❗5            │" Hidden by multi-width symbols: [(2, " "), (25, " ")]
"│#tag #tag2                            │"
"│┌❌ subtask with another long title ─┐│" Hidden by multi-width symbols: [(3, " ")]
"││that should wrap around             ││"
"││test                                ││"
"││desc                                ││"
"│└────────────────────────────────────┘│"
"│┌❌ subtask test─────────────────────┐│" Hidden by multi-width symbols: [(3, " ")]
"││#tag #tag2                          ││"
"│└────────────────────────────────────┘│"
"│┌❌ subtask test with a long title ──┐│" Hidden by multi-width symbols: [(3, " ")]
"││123456789 1 2 3                     ││"
"││📅 2016/07/08 09:10:11 ❗5          ││" Hidden by multi-width symbols: [(3, " "), (26, " ")]
"││test                                ││"
"││desc                                ││"
"│└────────────────────────────────────┘│"
"│                                      │"
"│                                      │"
"│                                      │"
"│                                      │"
"│                                      │"
"│                                      │"
"│                                      │"
"│                                      │"
"│                                      │"
"│                                      │"
"│                                      │"
"│                                      │"
"│                                      │"
"│                                      │"
"│                                      │"
"│                                      │"
"│                                      │"
"│                                      │"
"│                                      │"
"│                                      │"
"│                                      │"
"└──────────────────────────────────────┘"
